org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/ReadRequest.java

```diff
@@ -21,33 +22,58 @@
 public class ReadRequest<V> extends AbstractRequest<V> {
 	private static final Logger LOG = Logger.getLogger(ReadRequest.class);
 
-	private final Function1<? super CancelIndicator, ? extends V> cancellable;
+	private final Function1<? super CancelIndicator, ? extends V> readOperation;
+
+	/**
+	 * The initializer future allows to track the running state of this request, e.g. if it was already started or not.
+	 */
+	private final CompletableFuture<Void> initializer;
 
 	private final ExecutorService executor;
 
-	public ReadRequest(RequestManager requestManager, Function1<? super CancelIndicator, ? extends V> cancellable,
+	public ReadRequest(RequestManager requestManager, Function1<? super CancelIndicator, ? extends V> readOperation,
 			ExecutorService executor) {
 		super(requestManager);
-		this.cancellable = cancellable;
+		this.readOperation = readOperation;
 		this.executor = executor;
+		this.initializer = new CompletableFuture<>();
+		this.initializer.thenRun(this::doRun);
 	}
-
+	
+	@Override
+	protected void cancel(boolean mayInterruptIfRunning) {
+		super.cancel(mayInterruptIfRunning);
+		if (initializer.cancel(mayInterruptIfRunning)) {
+			cancelResult(mayInterruptIfRunning);
+		}
+	}
+	
 	@Override
 	public void run() {
-		if (result.isCancelled()) {
+		initializer.complete(null);
+	}
+
+	private void doRun() {
+		if (isDone()) {
 			return;
 		}
 		this.executor.submit(() -> {
 			try {
+				if (isDone()) {
+					return;
+				}
 				cancelIndicator.checkCanceled();
-				result.complete(cancellable.apply(cancelIndicator));
+				V readResult = readOperation.apply(cancelIndicator);
+				complete(readResult);
```

Comment:
I'm not sure whether I understand it all but I think you are attempting to make sure that consumers of the `readResult` are canceled if the `ReadRequest` itself is canceled.  Suppose that a cancellation is requested after `readResult` is produced.  Then `complete` will go ahead  regardless since `org.eclipse.xtext.ide.server.concurrent.AbstractRequest.cancel()` doesn't cancel the `ResultFuture`?   Is that intended behaviour?

Reply:
The line of thought was like this:
Consumers of the readResult must wait until the read request is finished (normally or exceptionally), to ensure that we do not start a subsequent write request when there are still readers. Now that the readOperation did complete normally (let's say it ignored the cancelIndicator), it was considered to be ok to return the result independently from intermitted cancel attempts. Does that make sense or am I in weird state of mind?